content/tutorials/portenta-h7/por-ard-flash/content.md

```diff
@@ -0,0 +1,256 @@
+
+```cpp
+/**
+   Helper functions for calculating limits for the FlashIAP block device
+ * */
+
+#pragma once
+
+#include <Arduino.h>
+#include <FlashIAP.h>
+#include <FlashIAPBlockDevice.h>
+
+using namespace mbed;
+
+// An helper struct for FlashIAP limits
+struct FlashIAPLimits {
+  size_t flash_size;
+  uint32_t start_address;
+  uint32_t available_size;
+};
+
+// Get the actual start address and available size for the FlashIAP Block Device
+// considering the space already occupied by the sketch.
+FlashIAPLimits getFlashIAPLimits()
+{
+
+  auto align_down = [](uint64_t val, uint64_t size) {
+    return (((val) / size)) * size;
+  };
+  auto align_up = [](uint32_t val, uint32_t size) {
+    return (((val - 1) / size) + 1) * size;
+  };
+
+  size_t flash_size;
+  uint32_t flash_start_address;
+  uint32_t start_address;
+  FlashIAP flash;
+
+  auto result = flash.init();
+  if (result != 0)
+    return { };
+
+  // Find the start of first sector after text area
+  int sector_size = flash.get_sector_size(FLASHIAP_APP_ROM_END_ADDR);
+  start_address = align_up(FLASHIAP_APP_ROM_END_ADDR, sector_size);
+  flash_start_address = flash.get_flash_start();
+  flash_size = flash.get_flash_size();
+
+  result = flash.deinit();
+
+  int available_size = flash_start_address + flash_size - start_address;
+  if (available_size % (sector_size * 2)) {
+    available_size = align_down(available_size, sector_size * 2);
+  }
+
+  return { flash_size, start_address, available_size };
+}
+```
+
+## 4. Add the Flash IAP Libaries 
+
+Get a few helpers from the supporting header file:
+
+```
+#include "FlashIAPLimits.h"
+```
+
+This header file defines the `getFlashIAPLimits` function that will calculate the starting point and the size of the flash storage that is available on the flash.
+
+```
+auto [flash_size, start_address, iap_size] = getFlashIAPLimits();
+```
+
+## 5. Intialise the Block Device 
+
+Let's create the block device using the calculated limits. Initialize the block device before reading or write data.
+
+```
+FlashIAPBlockDevice blockDevice(start_address, iap_size);
+blockDevice.init();
+```
+
+Remember to always **allocate** a **flash-erase-size-wide** storage area to read and write data from the flash:
+
+```
+const size_t size { blockDevice.get_erase_size() };
+char buffer[size] {}; 
+blockDevice.read(buffer, 0, size);
+```
+
+Don't forget to **erase** the flash block **before** being able to **program** it with your data, i.e. to write data on it (that's the way flash memories work):
+
+```
+ blockDevice.erase(0, size);
+ blockDevice.program(buffer, 0, size);
+```
+
+*De-init* the block-device when done:
+
+```
+blockDevice.deinit(); 
+```
+
+Please, remember that **the data stored** on the flash memory will be **erased** at every **sketch upload** and will only be **retained** through successive sketch executions, e.g. after **power cycling** or **resetting** the board.
```

Comment:
```suggestion
Please, remember that **the data stored** on the flash memory will be **erased** at every **sketch upload** and will only be **retained** through successive sketch executions, e.g. after **power cycling** or **resetting** the board.
```
[note]

**Note :** **The data stored** on the flash memory will be **erased** at every **sketch upload** and will only be **retained** only through successive sketch executions, e.g. after **power cycling** or **resetting** the board.

[/note]